File: fsharp/fox-goose-bag-of-corn/test/puzzle-test.fs
module puzzle_test

open Expecto
open fox_goose_bag_of_corn.puzzle

[<Tests>]
let tests =
    let crossing_plan = river_crossing_plan()
    let left_bank =
        crossing_plan
        |> List.map (fun steps -> steps |> List.head)
    let right_bank =
        crossing_plan
        |> List.map (fun steps -> steps |> List.last)
    let boat_positions =
        crossing_plan
        |> List.map (fun steps -> steps |> List.item 1)
    testList "puzzle" [
        testCase "you begin with the fox, goose and corn on one side of the river" <| fun _ ->
            Expect.containsAll (crossing_plan |> List.head) ([[Fox; Goose; Corn; You]; [Boat]; []]) ""

        testCase "you end with the fox, goose and corn on one side of the river" <| fun _ ->
            Expect.containsAll (crossing_plan |> List.last) ([[]; [Boat]; [You; Fox; Goose; Corn]]) ""

        testCase "the fox and the goose should never be left alone together" <| fun _ ->
            Expect.isEmpty
                ((left_bank @ right_bank)
                |> List.filter (fun bank -> (bank |> Set.ofList) = ([Fox; Goose] |> Set.ofList))) ""

        testCase "the goose and the corn should never be left alone together" <| fun _ ->
            Expect.isEmpty
                ((left_bank @ right_bank)
                |> List.filter (fun bank -> (bank |> Set.ofList) = ([Goose; Corn] |> Set.ofList))) ""

        testCase "the boat can carry only you plus one other" <| fun _ ->
            Expect.isEmpty
                (boat_positions |> List.filter (fun boat -> boat.Length > 3)) ""

        testCase "moves are valid" <| fun _ ->
            let lft_moves = left_bank
            let mid_moves = boat_positions
            let rht_moves = right_bank

            let validate_move (step1 : Puzzle list) (step2 : Puzzle list) =
                let diff1 =
                    step1
                    |> List.filter (fun n -> not (step2 |> List.contains n))
                    |> List.distinct
                let diff2 =
                    step2
                    |> List.filter (fun n -> not (step1 |> List.contains n))
                    |> List.distinct
                let diffs = diff1 @ diff2
                let diff_num = diffs |> List.length
                Expect.isGreaterThan 3 diff_num ""
                Expect.contains diffs You ""
                step2

            lft_moves |> List.reduce validate_move |> ignore
            mid_moves |> List.reduce validate_move |> ignore
            rht_moves |> List.reduce validate_move |> ignore
    ]

[<EntryPoint>]
let main argv =
    runTestsInAssembly defaultConfig argv

File: csharp/fox-goose-bag-of-corn/test/puzzle-test.cs
namespace fox_goose_bag_of_corn;

using Xunit;
using System.Linq;
using System.Collections.Generic;
using static fox_goose_bag_of_corn.puzzle;

public class puzzle_test {

    static List<List<List<P>>> crossing_plan = river_crossing_plan();
    List<List<P>> left_bank =
        crossing_plan
        .Select(n => n.First())
        .ToList();
    List<List<P>> right_bank =
        crossing_plan
        .Select(n => n.Last())
        .ToList();
    List<List<P>> boat_positions =
        crossing_plan
        .Select(n => n.ElementAt(1))
        .ToList();

    [Fact]
    public void begin_fox_goose_corn_on_left_side_of_river() {
        Assert.Equal (
            new List<List<P>>() {
                new List<P>() { P.Fox, P.Goose, P.Corn, P.You },
                new List<P>() { P.Boat },
                new List<P>() {}
            },
            crossing_plan.First()
        );
    }

    [Fact]
    public void end_fox_goose_corn_on_right_side_of_river() {
        Assert.Equal (
            new List<List<P>>() {
                new List<P>() { },
                new List<P>() { P.Boat },
                new List<P>() { P.You, P.Fox, P.Goose, P.Corn }
            },
            crossing_plan.Last()
        );
    }

    [Fact]
    public void fox_and_goose_never_left_alone_together() {
        Assert.Empty (
            left_bank.Concat(right_bank)
            .Where(bank => bank.Distinct() == (new List<P> { P.Fox, P.Goose}).Distinct())
        );
    }

    [Fact]
    public void goose_and_corn_never_left_alone_together() {
        Assert.Empty (
            left_bank.Concat(right_bank)
            .Where(bank => bank.Distinct() == (new List<P> { P.Goose, P.Corn }).Distinct())
        );
    }

    [Fact]
    public void boat_only_carry_you_plus_one_other() {
        Assert.Empty (
            boat_positions
            .Where(boat => boat.Count() > 3)
        );
    }

    [Fact]
    public void moves_are_valid() {
        var lft_moves = left_bank;
        var mid_moves = boat_positions;
        var rht_moves = right_bank;

        List<P> validate_move(List<P> step1, List<P> step2) {
            var diff1 =
                step1
                .Where(n => !(step2.Contains(n)))
                .Distinct();
            var diff2 =
                step2
                .Where(n => !(step1.Contains(n)))
                .Distinct();
            var diffs = diff1.Concat(diff2);
            var diff_num = diffs.Count();
            Assert.True(diff_num > 3);
            Assert.Contains(P.You, diffs);
            return step2;
        }

        lft_moves.Aggregate(validate_move);
        mid_moves.Aggregate(validate_move);
        rht_moves.Aggregate(validate_move);
    }
}
